zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -142,13 +154,16 @@ impl Chain {
             self.spent_utxos == other.spent_utxos &&
 
             // note commitment trees
+            self.sprout_note_commitment_tree.root() == other.sprout_note_commitment_tree.root() &&
```

Comment:
This line causes the test `forked_equals_pushed_genesis` to fail here https://github.com/ZcashFoundation/zebra/blob/0fafa30e5d19c10e7f20085ab371ad9834274875/zebra-state/src/service/non_finalized_state/tests/prop.rs#L189

Reply:
`forked_equals_pushed_genesis` checks that you get the same chain if you:
* forking the original chain, then push some blocks, or
* push the same blocks to the original chain.

It looks like sprout anchors are implemented the same in both the fork and push code.
* Push: https://github.com/ZcashFoundation/zebra/blob/main/zebra-state/src/service/non_finalized_state.rs#L128
* Fork: https://github.com/ZcashFoundation/zebra/blob/main/zebra-state/src/service/non_finalized_state.rs#L167

So I'm wondering if there's a bug in the `NoteCommitmentTree`, `Transaction::sprout_note_commitments`, or the serialization. It's the first time a lot of that code has been tested.

You could try printing the list of block heights/hashes, transaction indexes/hashes, or note commitments that are added to `self` and `other`. That might help you find what's different.

You can stop the proptests running extra shrink tests using:
```sh
export PROPTEST_MAX_SHRINK_ITERS=0
```
That might help reduce the number of logs you have to look through.

Reply:
Note that `eq_internal_state` checks a lot of stuff and there isn't an obvious way to tell which one failed. It could be the note commitment tree, the anchors, or the nullifiers (or some regression elsewhere).

By commenting out the Sprout checks and re-enabling one at a a time (very advanced debugging technique :laughing: ) it seems that both note commitment trees and anchors are not matching. The nullifier check is passing.

On forks, trees are recreated from the finalized state tip (in this test, it uses the `default()` of the tree for that). So I'm not sure what's going on, I second @teor2345 suggestion of printing what is being pushed to the tree and comparing to see where the difference is.

Reply:
It might also help to assert equality after every transaction or block is added to the chain state. That way, the test will panic on the first mismatch.

This might require rewriting the test.

Can you reproduce this issue using a test that:
1. Adds 1 block
2. Forks the chain
3. Adds 1 more block to both chains

If you can, the test would be a lot faster, and you could add transactions from the second blocks one by one.